staking/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+                self.tokens_per_stake = self
+                    .tokens_per_stake
+                    .saturating_add((produced_new * DECIMALS_COUNT) / self.total_staked);
+            }
+
+            self.reward_produced = self.reward_produced.saturating_add(produced_new);
+        }
+    }
+
+    /// Calculates the maximum possible reward
+    /// The reward that the depositor would have received if he had initially paid this amount
+    /// Arguments:
+    /// `amount`: the number of tokens
+    fn get_max_reward(&self, amount: u128) -> u128 {
+        (amount * self.tokens_per_stake) / DECIMALS_COUNT
+    }
+
+    /// Calculates the reward of the staker that is currently avaiable
+    fn calc_reward(&mut self) -> u128 {
+        if let Some(staker) = self.stakers.get(&msg::source()) {
+            return self.get_max_reward(staker.balance) + staker.reward_allowed
+                - staker.reward_debt
+                - staker.distributed;
+        }
+
+        panic!("calc_reward(): Staker {:?} not found", msg::source());
+    }
+
+    /// Sets the reward to be distributed within distribution time
+    /// param 'reward' The value of the distributed reward
+    fn set_reward_total(&mut self, reward: u128) {
+        if reward == 0 {
+            panic!("set_reward_total(): reward is null");
+        }
+
+        self.update_reward();
+        self.all_produced = self.reward_produced;
+        self.produced_time = exec::block_timestamp();
+        self.reward_total = reward;
+    }
+
+    /// Stakes the tokens
+    /// Arguments:
+    /// `amount`: the number of tokens for the stake
+    async fn stake(&mut self, amount: u128) {
+        if amount == 0 {
+            panic!("enter(): amount is null");
+        }
+
+        let token_address = self.staking_token_address;
+
+        transfer_tokens(&token_address, &msg::source(), &exec::program_id(), amount).await;
+
+        msg::reply(StakingEvent::StakeAccepted(amount), 0).unwrap();
+
+        self.update_reward();
+        let amount_per_token = self.get_max_reward(amount);
+
+        self.stakers
+            .entry(msg::source())
+            .and_modify(|stake| {
+                stake.reward_debt = stake.reward_debt.saturating_add(amount_per_token);
+                stake.balance = stake.balance.saturating_add(amount);
+            })
+            .or_insert_with(|| Staker {
+                reward_debt: amount_per_token,
+                balance: amount,
+                ..Default::default()
+            });
+
+        self.total_staked = self.total_staked.saturating_add(amount);
+    }
+
+    ///Sends reward to the staker
+    async fn send_reward(&mut self) {
+        self.update_reward();
+        let reward = self.calc_reward();
+
+        if reward == 0 {
+            panic!("send_reward(): reward is null");
+        }
+
+        let token_address = self.reward_token_address;
+
+        transfer_tokens(&token_address, &exec::program_id(), &msg::source(), reward).await;
```

Comment:
The external function is called using a module.

https://doc.rust-lang.org/stable/book/ch07-04-bringing-paths-into-scope-with-the-use-keyword.html?highlight=idiomatic#creating-idiomatic-use-paths

Anyway, I'd make this function inside `impl Staking` if it is not used outside.

Reply:
We discussed this function(`transfer_tokens()`) in Telegram. Returned it to the `lib.rs` module, but outside the `impl Staking`